### PR TITLE
[ci] Migrate from CODEOWNERS to codemention

### PR DESCRIPTION
**Pull request by wschurman**
# Why

Ref: https://exponent-internal.slack.com/archives/C1QP38NQ5/p1726788903508149

Closes ENG-13562.

Quoting from the universe repo equivalent https://app.graphite.dev/github/pr/expo/universe/15840/Remove-CODEOWNERS-replaced-with-Graphite-notifications:

> The main purpose of code review at Expo is for two or more people to be aware of how each part of the codebase works. Code review is a form of communication and collaboration, and part of the code review process is for authors to select reviewers. Additionally, we find the ideal number of reviewers to be one or two people on most PRs since it makes clear who is responsible for reviewing the code.
> 
> GitHub's implementation of CODEOWNERS is misaligned with these principles. @wschurman posited that CODEOWNERS may be poisonous towards a good engineering culture since it feels impersonal and gives the impression that code review is not the author's problem. CODEOWNERS is also overly rigid in that it forcibly adds all matching users as reviewers. This policing aspect of CODEOWNERS is not how we want to do engineering at Expo in most cases, except where we have high security requirements around access control.
> 
> However, we do want engineers to be accountable for different parts of the codebase and they need to be aware of changes to those parts. That is why we used to use CODEOWNERS. If GitHub were to introduce the concept of "Subscribers" on PRs, we'd happily use that feature instead.

We dogfooded this subscription model in universe, first with graphite subscriptions, then with codenotify (https://app.graphite.dev/github/pr/expo/universe/16210/universe-Add-codenotify), then finally switched to codemention (https://app.graphite.dev/github/pr/expo/universe/16225/universe-Switch-from-codenotify-to-codemention).

So far codemention has achieved this goal in universe. So this PR does the same thing for expo/expo.

Some additional notes about CODEOWNERS from discussions:
- Too many assigned reviewers can result in slower code review since then review is “somebody else’s problem”/“I’ll wait for one of the other reviewers before I review”. When everyone has this gut reaction, code review doesn’t occur.
- CODEOWNERS in this repo adds too many reviewers and it can feel socially awkward to remove them, even if you have only one or two reviewers in mind for the primary review responsibility. This is especially true for new team members who feel like they don't have the authority to select reviewers of their own PR, which is something we want to explicitly counter. An author should have the ability to select their reviewers.
- No way to express primary review responsibility in CODEOWNERS. Adding a code review request should mean, “I expect you as a reviewer of this to do a review”.

# How

Manually convert CODEOWNERS file to codemention.yaml.

Originally I started working on a way for codemention to read the CODEOWNERS file but the syntax differences are probably too confusing: https://github.com/tobyhs/codemention/pull/14/

# Test Plan

Tested by stacking a PR on top of this one that touches a file, see that codemention comment is posted: https://app.graphite.dev/github/expo/expo/pull/31644

Also as a sanity check, I put it through a yaml validator to ensure it's valid.

# Checklist

- [ ] Documentation is up to date to reflect these changes (eg: https://docs.expo.dev and README.md).
- [ ] Conforms with the [Documentation Writing Style Guide](https://github.com/expo/expo/blob/main/guides/Expo%20Documentation%20Writing%20Style%20Guide.md)
- [ ] This diff will work correctly for `npx expo prebuild` & EAS Build (eg: updated a module plugin).

### CODE REVIEW

**Comment by wschurman** [comment]
* **#31639** <a href="https://app.graphite.dev/github/pr/expo/expo/31639?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @wschurman and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: .github/codemention.yml [hunk 1]
```diff
@@ -0,0 +1,232 @@
+  - patterns: ['.github/workflows/**']
```
Comment (tobyhs) [inline .github/codemention.yml:1]: You need a root/top level `rules` key.

Comment (wschurman) [reply .github/codemention.yml:2]: Good call. Awesome work on the codemention action!

**Comment by tobyhs** [comment]
> This is hard to test since it is a `pull_request_target` github action, which only runs from main. So it needs to be checked in, and then a test PR needs to be put up.

I think you can test this before merging to main by creating a new branch based off of your PR branch (`@wschurman/09-22-_ci_migrate_from_codeowners_to_codemention`), then open a PR for that new branch that targets your PR branch (I believe I used this technique to test codemention on a repo at a previous company I worked at).

**Comment by wschurman** [comment]
> > This is hard to test since it is a `pull_request_target` github action, which only runs from main. So it needs to be checked in, and then a test PR needs to be put up.
> 
> I think you can test this before merging to main by creating a new branch based off of your PR branch (`@wschurman/09-22-_ci_migrate_from_codeowners_to_codemention`), then open a PR for that new branch that targets your PR branch (I believe I used this technique to test codemention on a repo at a previous company I worked at).

Good to know! I'll give it a try.

**Review by brentvatne — APPROVED** [review]

**Review by Kudo — APPROVED** [review]

**Pull request merged**